metadata/lib.rs

```diff
@@ -0,0 +1,418 @@
+}
+
+impl SourceLanguage {
+    /// Constructs a new SourceLanguage.
+    pub fn new(language: Language, version: Version) -> Self {
+        SourceLanguage { language, version }
+    }
+}
+
+impl Serialize for SourceLanguage {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl Display for SourceLanguage {
+    fn fmt(&self, f: &mut Formatter<'_>) -> DisplayResult {
+        write!(f, "{} {}", self.language, self.version)
+    }
+}
+
+/// The language in which the smart contract is written.
+#[derive(Debug)]
+pub enum Language {
+    Ink,
+    Solidity,
+    AssemblyScript,
+}
+
+impl Display for Language {
+    fn fmt(&self, f: &mut Formatter<'_>) -> DisplayResult {
+        match self {
+            Self::Ink => write!(f, "ink!"),
+            Self::Solidity => write!(f, "Solidity"),
+            Self::AssemblyScript => write!(f, "AssemblyScript"),
+        }
+    }
+}
+
+/// A compiler used to compile a smart contract.
+#[derive(Debug)]
+pub struct SourceCompiler {
+    compiler: Compiler,
+    version: Version,
+}
+
+impl Display for SourceCompiler {
+    fn fmt(&self, f: &mut Formatter<'_>) -> DisplayResult {
+        write!(f, "{} {}", self.compiler, self.version)
+    }
+}
+
+impl Serialize for SourceCompiler {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl SourceCompiler {
+    pub fn new(compiler: Compiler, version: Version) -> Self {
+        SourceCompiler { compiler, version }
+    }
+}
+
+/// Compilers used to compile a smart contract.
+#[derive(Debug, Serialize)]
+pub enum Compiler {
+    RustC,
+    Solang,
+}
+
+impl Display for Compiler {
+    fn fmt(&self, f: &mut Formatter<'_>) -> DisplayResult {
+        match self {
+            Self::RustC => write!(f, "rustc"),
+            Self::Solang => write!(f, "solang"),
+        }
+    }
+}
+
+/// Metadata about a smart contract.
+#[derive(Debug, Serialize)]
+pub struct Contract {
+    name: String,
+    version: Version,
+    authors: Vec<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    description: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    documentation: Option<Url>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    repository: Option<Url>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    homepage: Option<Url>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    license: Option<String>,
+}
+
+impl Contract {
+    /// Constructs a new Contract.
+    pub fn new(
+        name: String,
+        version: Version,
+        authors: Vec<String>,
+        description: Option<String>,
+        documentation: Option<Url>,
+        repository: Option<Url>,
+        homepage: Option<Url>,
+        license: Option<String>,
+    ) -> Self {
```

Comment:
Personally I'd prefer a builder pattern. This would be especially handy for all these optional parameters. Doesn't need to be a fancy build pattern with compile-time checks in this case.